llama_index/vector_stores/cassandra.py

```diff
@@ -0,0 +1,248 @@
+"""Cassandra / Astra DB Vector store index.
+
+An index based on a DB table with vector search capabilities,
+powered by the cassIO library
+
+"""
+
+import logging
+from typing import Any, List, Optional, cast
+
+
+from llama_index.vector_stores.utils import (
+    DEFAULT_TEXT_KEY,
+    metadata_dict_to_node,
+    node_to_metadata_dict,
+)
+from llama_index.indices.query.embedding_utils import (
+    get_top_k_mmr_embeddings,
+)
+
+from llama_index.vector_stores.types import (
+    # MetadataFilters,
+    NodeWithEmbedding,
+    VectorStore,
+    VectorStoreQuery,
+    VectorStoreQueryMode,
+    VectorStoreQueryResult,
+)
+
+_logger = logging.getLogger(__name__)
+
+DEFAULT_MMR_PREFETCH_FACTOR = 4.0
+
+
+class CassandraVectorStore(VectorStore):
+    """Cassandra Vector Store.
+
+    An abstraction of a Cassandra table with
+    vector-similarity-search. Documents, and their embeddings, are stored
+    in a Cassandra table and a vector-capable index is used for searches.
+    The table does not need to exist beforehand: if necessary it will
+    be created behind the scenes.
+
+    All Cassandra operations are done through the cassIO library.
+
+    Args:
+        session (cassandra.cluster.Session): the Cassandra session to use
+        keyspace (str): name of the Cassandra keyspace to work in
+        table (str): table name to use. If not existing, it will be created.
+        embedding_dimension (int): length of the embedding vectors in use.
+        ttl_seconds (Optional[int]): expiration time for inserted entries.
+            Default is no expiration.
+
+    """
+
+    stores_text: bool = True
+    flat_metadata: bool = True
+
+    def __init__(
+        self,
+        session: Any,
+        keyspace: str,
+        table: str,
+        embedding_dimension: int,
+        ttl_seconds: Optional[int] = None,
+    ) -> None:
+        import_err_msg = "`cassio` package not found, please run `pip install cassio`"
+        try:
+            import cassio  # noqa: F401
+        except ImportError:
+            raise ImportError(import_err_msg)
+
+        self._session = session
+        self._keyspace = keyspace
+        self._table = table
+        self._embedding_dimension = embedding_dimension
+        self._ttl_seconds = ttl_seconds
+
+        _logger.debug("Creating VectorTable")
+        self.vector_table = cassio.vector.VectorTable(
+            session=self._session,
+            keyspace=self._keyspace,
+            table=self._table,
+            embedding_dimension=self._embedding_dimension,
+            primary_key_type="TEXT",
+        )
+
+    def add(
+        self,
+        embedding_results: List[NodeWithEmbedding],
+    ) -> List[str]:
+        """Add embedding results to index.
+
+        Args
+            embedding_results: List[NodeWithEmbedding]: list of embedding results
+
+        """
+        (document_ids0, documents, document_metadatas, embeddings,) = zip(
```

Comment:
Just for the sake of code clarity, while clever, this is kind of a confusing way to build these lists. But, up to you.

Also, is the word document specific to Cassandra? I would prefer to use the term `node` instead for these variable names, as documents vs. nodes mean something specific in the llama index code

Reply:
In Cassandra world, generally we're dealing with partitions and rows similar to a relational database, but in the context of this use case, in the nosql world in general, a document is used to define some structure which has hierarchy. As far conforming to LlamaIndex terminology, I don't think it makes much of a difference. @hemidactylus  what do you think.

Reply:
I just adapted the variable names to refer to `nodes` and reduce the confusion (up to where the `self.vector_table` methods are called, which currently expect a "document-based metaphor").